src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs

```diff
@@ -242,10 +267,40 @@ public void Tick(DateTimeOffset now)
 
                 if (_timeoutAction == TimeoutAction.SendTimeoutResponse)
                 {
+                    SetTimeoutResponse();
+                }
+
+                Timeout();
+            }
+            else if (_timingReads)
+            {
+                _readTimingElapsed += timestamp - _lastTimestamp;
+
+                if (_readTimingElapsed > _frame.RequestBodyTimeout.Ticks)
```

Comment:
You don't have to do the math. You could just limit one of rate/timeout/size and set the others unlimited if that's want you want.